src/main/java/org/zalando/nakadi/service/subscription/StreamingContext.java

```diff
@@ -265,18 +267,23 @@ private void rebalance() {
         if (null != sessionListSubscription) {
             // This call is needed to renew subscription for session list changes.
             final List<String> newSessions = sessionListSubscription.getData();
-            final String newHash = ZkSubscriptionClient.Topology.calculateSessionsHash(newSessions);
+            final String sessionsHash = ZkSubscriptionClient.Topology.calculateSessionsHash(newSessions);
             zkClient.runLocked(() -> {
                 final ZkSubscriptionClient.Topology topology = zkClient.getTopology();
 
-                if (!topology.isSameHash(newHash)) {
-                    log.info("Performing rebalance, hash changed: {}", newHash);
-                    final Partition[] changeset = rebalancer.apply(zkClient.listSessions(), topology.getPartitions());
+                if (!topology.isSameHash(sessionsHash)) {
+                    log.info("Performing rebalance, hash changed: {}", sessionsHash);
+                    final Collection<Session> newSessionsUnderLock = zkClient.listSessions();
+
+                    // after taking the lock list of sessions may change, so we need to update hash to correct value.
+                    final String actualHash = ZkSubscriptionClient.Topology.calculateSessionsHash(
```

Comment:
I think it is better to move calculation of the hash in the `if` below

Reply:
done